In [2]:
import tensorflow as tf
print("Num GPUs Available: ", len(tf.config.list_physical_devices('GPU')))

Num GPUs Available:  1


In [10]:
import gpflow
from gpflow.kernels import Matern52 as mt52
import numpy as np
import time
from concurrent.futures import ThreadPoolExecutor

n_samples, n_iters = 100,100
n_data_points = 200

start = time.time()
X_train = np.random.rand(n_data_points,2)
y_train = np.random.rand(n_data_points,1)

def gpflow_sample(sample):
    variance = sample[:1][0]
    lengths = (sample[1:])
    kern = mt52(variance=variance, lengthscales=lengths)
    model = gpflow.models.GPR(
        (X_train, y_train),
        kernel=kern)

    mean, cov = model.predict_y(X_test, full_cov=False)
    mean = np.squeeze(mean.numpy())
    cov = np.squeeze(cov.numpy())

    return 0

start = time.time


for i in range(n_iters):
    X_test = np.random.rand(5,2)
    samples = np.random.rand(n_samples, 3)
    futures = [gpflow_sample(sample) for sample in samples]

    print(f'iter {i} done')

end = time.time()

print(f'{n_samples* n_iters} trained equivalent to a {n_samples} sample, {n_iters} iteration smc sampler, this was done in {end-start:.4f} seconds')

iter 0 done
iter 1 done
iter 2 done
iter 3 done
iter 4 done
iter 5 done
iter 6 done
iter 7 done
iter 8 done
iter 9 done
iter 10 done
iter 11 done
iter 12 done
iter 13 done
iter 14 done
iter 15 done
iter 16 done
iter 17 done
iter 18 done
iter 19 done
iter 20 done
iter 21 done
iter 22 done
iter 23 done
iter 24 done
iter 25 done
iter 26 done
iter 27 done
iter 28 done
iter 29 done
iter 30 done
iter 31 done
iter 32 done
iter 33 done
iter 34 done
iter 35 done
iter 36 done
iter 37 done
iter 38 done
iter 39 done
iter 40 done
iter 41 done
iter 42 done
iter 43 done
iter 44 done
iter 45 done
iter 46 done
iter 47 done
iter 48 done
iter 49 done
iter 50 done
iter 51 done
iter 52 done
iter 53 done
iter 54 done
iter 55 done
iter 56 done
iter 57 done
iter 58 done
iter 59 done
iter 60 done
iter 61 done
iter 62 done
iter 63 done
iter 64 done
iter 65 done
iter 66 done
iter 67 done
iter 68 done
iter 69 done
iter 70 done
iter 71 done
iter 72 done
iter 73 done
iter 74 done
iter 75 done
iter 76 done
iter 77 d

TypeError: unsupported operand type(s) for -: 'float' and 'builtin_function_or_method'

it's definitely using gpu vram, but only 12% of the gpu might be able to multithread just point a join condition in each for loop, also doesn't seem to offloat the meme usage 17% at 200 data points instead of 20 - but basically the same speed for each iter, really just need to write a better proposal - also am curious to see the dif on normal windows vs - me trying to get it working on ubuntu with gpu usage - which i think is curr being used bc its quite hot (60C) idk tho


outputs:
* 10000 trained equivalent to a 100 sample, 100 iteration smc sampler, this was done in 323.0208 seconds - with GPU I think on 20 datapoints 5 test points
* 10000 trained equivalent to a 100 sample, 100 iteration smc sampler, this was done in 350 seconds - with GPU I think on 200 datapoints 5 test points